R/as_draws_df.R

```diff
@@ -61,58 +61,50 @@ as_draws_df.draws_df <- function(x, ...) {
 #' @rdname draws_df
 #' @export
 as_draws_df.draws_matrix <- function(x, ...) {
-  class(x) <- "matrix"
-  draws <- as.integer(rownames(x))
-  rownames(x) <- NULL
+  if (ndraws(x) == 0L) {
+    return(empty_draws_df(variables(x)))
+  }
   x <- tibble::as_tibble(x)
-  x$.chain <- rep(1L, nrow(x))
-  x$.iteration <- draws
-  x$.draw <- draws
+  x[".chain"] <- 1L
+  x[c(".iteration", ".draw")] <- 1L:nrow(x)
   class(x) <- class_draws_df()
   x
 }
 
 #' @rdname draws_df
 #' @export
 as_draws_df.draws_array <- function(x, ...) {
-  if (ndraws(x) == 0) {
+  if (ndraws(x) == 0L) {
     return(empty_draws_df(variables(x)))
   }
   iteration_ids <- iteration_ids(x)
   chain_ids <- chain_ids(x)
-  rownames(x) <- NULL
-  out <- named_list(chain_ids)
-  for (i in seq_along(out)) {
-    out[[i]] <- drop_dims_or_classes(x[, i, ], dims = 2, reset_class = TRUE)
-    class(out[[i]]) <- "matrix"
-    out[[i]] <- tibble::as_tibble(out[[i]])
-    out[[i]]$.chain <- chain_ids[i]
-    out[[i]]$.iteration <- iteration_ids
-    out[[i]]$.draw <- compute_draw_ids(chain_ids[i], iteration_ids)
-  }
-  out <- do.call(rbind, out)
-  class(out) <- class_draws_df()
-  out
+  x <- apply(x, 3L, c)
+  x <- tibble::as_tibble(x)
+  x[".chain"] <- rep(chain_ids, each = max(iteration_ids))
+  x[".iteration"] <- rep(iteration_ids, max(chain_ids))
+  x[".draw"] <- 1L:nrow(x)
+  class(x) <- class_draws_df()
+  x
 }
 
 #' @rdname draws_df
 #' @export
 as_draws_df.draws_list <- function(x, ...) {
-  if (ndraws(x) == 0) {
+  if (ndraws(x) == 0L) {
     return(empty_draws_df(variables(x)))
   }
   iteration_ids <- iteration_ids(x)
   chain_ids <- chain_ids(x)
-  out <- named_list(chain_ids)
-  for (i in seq_along(out)) {
-    out[[i]] <- tibble::as_tibble(x[[i]])
-    out[[i]]$.chain <- chain_ids[i]
-    out[[i]]$.iteration <- iteration_ids
-    out[[i]]$.draw <- compute_draw_ids(chain_ids[i], iteration_ids)
-  }
-  out <- do.call(rbind, out)
-  class(out) <- class_draws_df()
-  out
+  vars <- names(x[[1L]])
+  x <- do.call(rbind.data.frame, x)
+  colnames(x) <- vars
+  x <- tibble::as_tibble(x)
+  x[".chain"] <- rep(chain_ids, each = max(iteration_ids))
+  x[".iteration"] <- rep(iteration_ids, max(chain_ids))
```

Comment:
Similarly, this vectorization might cause a problem only if the `draws_list` input is skipping a draw or has randomly ordered draws but I was not able to produce that specific case. In my attempts to create edge cases to break this function, chain/iteration/draw ids are getting repaired or I get errors indicating that such `draws_list` structures are not possible.

Reply:
Also looks fine for me.